### PR TITLE
AVRO-3285: Upgrade JavaCC plugin and library

### PR DESCRIPTION
**Pull request by opwvhk**
Also resolve IDE warnings for idl.jj, which was previously impossible.

Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3285
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [X] My PR ~adds the following unit tests __OR__~ does not need testing for this extremely good reason:  
  Upgrading plugins and libraries does not add functionality; existing tests must pass though

### Commits

- [X] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: lang/java/compiler/pom.xml [hunk 1]
```diff
@@ -24,7 +24,7 @@
     <artifactId>avro-parent</artifactId>
     <groupId>org.apache.avro</groupId>
     <version>1.12.0-SNAPSHOT</version>
-    <relativePath>../</relativePath>
+    <relativePath>../pom.xml</relativePath>
```
Comment (opwvhk) [inline lang/java/compiler/pom.xml:27]: Minor really, but the relative path should point to a pom.

File: lang/java/compiler/pom.xml [hunk 2]
```diff
@@ -133,11 +133,11 @@
               <classpathScope>test</classpathScope>
               <arguments>
                 <argument>-classpath</argument>
-                <classpath></classpath>
+                <classpath/>
                 <argument>org.apache.avro.compiler.specific.SchemaTask</argument>
                 <argument>${project.basedir}/src/test/resources/full_record_v1.avsc</argument>
                 <argument>${project.basedir}/src/test/resources/full_record_v2.avsc</argument>
-                <argument>${project.basedir}/target/generated-test-sources</argument>
+                <argument>${project.basedir}/target/generated-test-sources/javacc</argument>
```
Comment (opwvhk) [inline lang/java/compiler/pom.xml:140]: Generated code should be placed in a subdirectory per tool. I copied the directory name from line 178.

File: lang/java/compiler/pom.xml [hunk 3]
```diff
@@ -161,16 +159,14 @@
             </goals>
             <configuration>
               <sources>
-                <source>${project.basedir}/target/generated-test-sources</source>
+                <source>${project.basedir}/target/generated-test-sources/javacc</source>
```
Comment (opwvhk) [inline lang/java/compiler/pom.xml:162]: Mirror the change in line 140 to maintain a working build.

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj [hunk 4]
```diff
@@ -81,7 +81,7 @@ import org.apache.avro.util.internal.Accessor;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.node.*;
 
-import org.apache.commons.lang3.StringEscapeUtils;
+import org.apache.commons.text.StringEscapeUtils;
```
Comment (opwvhk) [inline lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:84]: `org.apache.commons.lang3.StringEscapeUtils` is deprecated in favor of `org.apache.commons.text.StringEscapeUtils`.

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj [hunk 5]
```diff
@@ -137,6 +137,7 @@ public class Idl implements Closeable {
     this.resourceLoader = parent.resourceLoader;
   }
 
+  @SuppressWarnings("RedundantThrows")
```
Comment (opwvhk) [inline lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:140]: A bit silly, but this is the one thing I cannot fix.

File: lang/java/pom.xml [hunk 6]
```diff
@@ -210,9 +211,16 @@
           </configuration>
         </plugin>
         <plugin>
-          <groupId>org.codehaus.mojo</groupId>
+          <groupId>org.javacc.plugin</groupId>
           <artifactId>javacc-maven-plugin</artifactId>
           <version>${javacc-plugin.version}</version>
+          <dependencies>
```
Comment (opwvhk) [inline lang/java/pom.xml:217]: The new plugin treats JavaCC as a provided dependency, so we need to add it explicitly.

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj [hunk 7]
```diff
@@ -1589,7 +1568,7 @@ Schema PrimitiveType():
 | "time_ms" { return LogicalTypes.timeMillis().addToSchema(Schema.create(Type.INT)); }
 | "timestamp_ms" { return LogicalTypes.timestampMillis().addToSchema(Schema.create(Type.LONG)); }
 | "local_timestamp_ms" { return LogicalTypes.localTimestampMillis().addToSchema(Schema.create(Type.LONG)); }
-| "decimal" { return DecimalTypeProperties(); }
+| "decimal" s = DecimalTypeProperties() { return s; }
```
Comment (martin-g) [inline lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:1571]: what is the benefit of this change ?

**Review by martin-g — APPROVED** [review]

**Review by RyanSkraba — APPROVED** [review]
LGTM too!

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj (hunk 7 above)
Comment (martin-g) [reply lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:1571]: @opwvhk Ping!

Comment (opwvhk) [reply lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:1571]: Nothing directly, but the rule DecimalTypeProperties no longer shows up as 'unused in IntelliJ.

**Comment by martin-g** [comment]
Thank you for the contribution, @opwvhk !

**Comment by martin-g** [comment]
I think this should not be backported to branch-1.11.
If I'm wrong then please let me know and I will cherry-pick it!

**Comment by opwvhk** [comment]
Correct @martin-g: it's an improvement, not a fix for a bug, so there's no need.

**Pull request merged**